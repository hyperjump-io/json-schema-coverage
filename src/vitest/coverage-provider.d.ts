import type { CoverageProviderModule } from "vitest";
import type { CustomProviderOptions } from "vitest/node";

export default CoverageProviderModule;

export type JsonSchemaCoverageProviderOptions = CustomProviderOptions & {
  provider: "custom";
  customProviderModule: "@hyperjump/json-schema-coverage/vitest-coverage-provider";
  include?: string[];
};
