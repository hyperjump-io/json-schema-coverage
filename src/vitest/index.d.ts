import type { SchemaObject } from "@hyperjump/json-schema";
import type { AsyncExpectationResult } from "@vitest/expect";
import "vitest";

declare module "vitest" {
  interface Matchers<R = unknown> {
    matchJsonSchema: (uriOrSchema: string | SchemaObject | boolean) => Promise<R>;
    toMatchJsonSchema: (uriOrSchema: string | SchemaObject | boolean) => Promise<R>;
  }
}

/**
 * Register a schema in your code base by it's path.
 *
 * _**NOTE**: This is **not** the same as the function from
 * [@hyperjump/json-schema](https://github.com/hyperjump-io/json-schema) that takes a schema._
 */
export const registerSchema: (filePath: string) => Promise<void>;

/**
 * Remove a registered schema in your code base by it's path.
 *
 * _**NOTE**: This is **not** the same as the function from
 * [@hyperjump/json-schema](https://github.com/hyperjump-io/json-schema) that takes the schema's `$id`._
 */
export const unregisterSchema: (filePath: string) => Promise<void>;

/**
 * A vitest matcher that can be used to validate a JSON-compatible value. It
 * can take a relative or full URI for a schema in your codebase. Use relative
 * URIs to reference a file and full URIs to reference the `$id` of a schema
 * you registered using the `registerSchema` function.
 *
 * You can use this matcher with an inline schema as well, but you will only
 * get coverage for schemas that are in files.
 *
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const matchJsonSchema: (instance: any, uriOrSchema: string | SchemaObject | boolean) => AsyncExpectationResult;
export const toMatchJsonSchema: typeof matchJsonSchema;

export { loadDialect, defineVocabulary, addKeyword } from "@hyperjump/json-schema/experimental";
