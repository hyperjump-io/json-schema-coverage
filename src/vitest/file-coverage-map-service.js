import { createHash } from "node:crypto";
import { existsSync } from "node:fs";
import { mkdir, readdir, readFile, rm, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { CoverageMapService } from "../coverage-map-service.js";

/**
 * @import { CoverageMapData } from "istanbul-lib-coverage"
 */

export class FileCoverageMapService extends CoverageMapService {
  /** @type string */
  #coverageMapsDirectory;

  /**
   * @param {string} coverageMapsDirectory
   */
  constructor(coverageMapsDirectory) {
    super();

    this.#coverageMapsDirectory = coverageMapsDirectory;
  }

  async open() {
    if (existsSync(this.#coverageMapsDirectory)) {
      await this.close();
    }

    await mkdir(this.#coverageMapsDirectory, { recursive: true });
  }

  async close() {
    await rm(this.#coverageMapsDirectory, {
      recursive: true,
      force: true,
      maxRetries: 10
    });
  }

  /** @type (coverageMapsDirectory: string) => Promise<FileCoverageMapService> */
  static async restoreFrom(coverageMapsDirectory) {
    const service = new FileCoverageMapService(coverageMapsDirectory);

    for (const file of await readdir(service.#coverageMapsDirectory, { withFileTypes: true })) {
      const path = resolve(file.parentPath, file.name);
      /** @type CoverageMapData */
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const coverageMap = JSON.parse(await readFile(path, "utf-8"));
      service.addCoverageMap(coverageMap);
    }

    return service;
  }

  /** @type (schemaPath: string) => Promise<string> */
  async addFromFile(schemaPath) {
    const schemaUri = await super.addFromFile(schemaPath);

    const coverageMap = this.getCoverageMap(schemaUri);
    const fileHash = createHash("md5").update(schemaUri).digest("hex");
    const coverageMapFile = resolve(this.#coverageMapsDirectory, fileHash);
    await writeFile(coverageMapFile, JSON.stringify(coverageMap));

    return schemaUri;
  }
}
