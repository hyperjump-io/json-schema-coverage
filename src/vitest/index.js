import "./json-schema-matcher.js";
