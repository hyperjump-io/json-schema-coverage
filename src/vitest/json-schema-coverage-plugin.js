/**
 * @import { Plugin } from "vite"
 * @import { VitestPluginContext, ResolvedCoverageOptions } from "vitest/node"
 */

import { resolve } from "path";

/** @type () => Plugin */
export const jsonSchemaCoveragePlugin = () => {
  return {
    name: "json-schema-coverage-plugin",

    /** @type (context: VitestPluginContext) => void */
    configureVitest(context) {
      const config = context.vitest.config;

      const coverage = /** @type ResolvedCoverageOptions<"custom"> */ (config.coverage);
      coverage.provider = "custom";
      coverage.customProviderModule = resolve(import.meta.dirname, "./coverage-provider.js");

      config.setupFiles.push(resolve(import.meta.dirname, "./register-matchers.js"));
    }
  };
};
