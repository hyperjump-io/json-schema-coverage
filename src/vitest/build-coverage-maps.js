import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { glob } from "tinyglobby";
import ignore from "ignore";
import { registerSchema } from "./index.js";
import { FileCoverageMapService } from "./file-coverage-map-service.js";

/**
 * @import { TestProject } from "vitest/node"
 */

/** @type (projects: TestProject) => Promise<void> */
export const setup = async (projects) => {
  const config = projects.vitest.config;

  const coverageService = new FileCoverageMapService(join(".json-schema-coverage", "maps"));

  // If --project filter is set pick only roots of resolved projects
  const roots = config.project?.length
    ? [...new Set(projects.vitest.projects.map((project) => project.config.root))]
    : [config.root];

  const include = "include" in config.coverage
    ? /** @type string[] */ (config.coverage.include)
    : ["**/*.schema.(json|yaml|yml)"];

  // Build coverage maps
  for (const root of roots) {
    const i = ignore();
    const gitignorePath = resolve(root, ".gitignore");
    if (existsSync(gitignorePath)) {
      const gitignore = await readFile(gitignorePath, "utf-8");
      i.add(gitignore);
    }

    let includedFiles = await glob(include, {
      cwd: root,
      dot: true,
      onlyFiles: true
    });

    const files = i
      .filter(includedFiles)
      .map((file) => resolve(root, file));

    for (const schemaPath of files) {
      try {
        await registerSchema(schemaPath);
      } catch (_error) {
      }
    }

    for (const file of files) {
      await coverageService.addFromFile(file);
    }
  }
};
