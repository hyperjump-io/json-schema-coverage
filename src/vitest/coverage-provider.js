import { existsSync, readdirSync } from "node:fs";
import * as fs from "node:fs/promises";
import path from "node:path";
import coverage from "istanbul-lib-coverage";
import libReport from "istanbul-lib-report";
import reports from "istanbul-reports";
import ignore from "ignore";
import { glob } from "tinyglobby";
import { resolve } from "pathe";
import c from "tinyrainbow";
import { coverageConfigDefaults } from "vitest/config";
import { registerSchema } from "./json-schema-matchers.js";
import { FileCoverageMapService } from "./file-coverage-map-service.js";

/**
 * @import {
 *   BaseCoverageOptions,
 *   CoverageProvider,
 *   CoverageProviderModule,
 *   ResolvedCoverageOptions,
 *   Vitest
 * } from "vitest/node"
 * @import { CoverageMap, CoverageMapData } from "istanbul-lib-coverage"
 */

/** @type CoverageProviderModule */
const JsonSchemaCoverageProviderModule = {
  /** @type CoverageProviderModule["getProvider"] */
  getProvider() {
    return new JsonSchemaCoverageProvider();
  }
};

/** @implements CoverageProvider */
class JsonSchemaCoverageProvider {
  name = "@hyperjump/json-schema-coverage/vitest/coverage-provider";

  ctx = /** @type Vitest */ ({});

  options = /** @type ResolvedCoverageOptions<"custom"> */ ({});

  /** @type Map<string, boolean> */
  globCache = new Map();

  coverageFilesDirectory = ".json-schema-coverage";
  coverageService = new FileCoverageMapService(".json-schema-coverage/maps");

  /** @type string[] */
  roots = [];

  include = ["**/*.schema.json", "**/*.schema.yaml", "**/*.schema.yml"];

  /** @type CoverageProvider["initialize"] */
  initialize(ctx) {
    this.ctx = ctx;

    const config = /** @type ResolvedCoverageOptions & { include: string[]; } */ (ctx.config.coverage);

    /** @type ResolvedCoverageOptions<"custom"> */
    this.options = {
      ...coverageConfigDefaults,

      // User's options
      ...config,

      // Resolved fields
      provider: "custom",
      customProviderModule: this.name,
      reportsDirectory: resolve(
        ctx.config.root,
        config.reportsDirectory || coverageConfigDefaults.reportsDirectory
      ),
      reporter: resolveCoverageReporters(config.reporter || coverageConfigDefaults.reporter)
    };

    // If --project filter is set pick only roots of resolved projects
    this.roots = ctx.config.project?.length
      ? [...new Set(ctx.projects.map((project) => project.config.root))]
      : [ctx.config.root];

    if ("include" in config) {
      this.include = config.include;
    }
  }

  /** @type CoverageProvider["resolveOptions"] */
  resolveOptions() {
    return /** @type NonNullable<any> */ (this.options);
  }

  /** @type CoverageProvider["clean"] */
  async clean(clean = true) {
    if (clean && existsSync(this.options.reportsDirectory)) {
      await fs.rm(this.options.reportsDirectory, {
        recursive: true,
        force: true,
        maxRetries: 10
      });
    }

    if (existsSync(this.coverageFilesDirectory)) {
      await fs.rm(this.coverageFilesDirectory, {
        recursive: true,
        force: true,
        maxRetries: 10
      });
    }

    await this.coverageService.open();

    await fs.mkdir(this.coverageFilesDirectory, { recursive: true });

    // Build coverage maps
    for (const root of this.roots) {
      let includedFiles = await glob(this.include, {
        cwd: root,
        dot: true,
        onlyFiles: true
      });
      const gitignorePath = path.resolve(root, ".gitignore");
      const gitignore = await fs.readFile(gitignorePath, "utf-8");
      const files = ignore()
        .add(gitignore)
        .filter(includedFiles)
        .map((file) => path.resolve(root, file));

      for (const schemaPath of files) {
        try {
          await registerSchema(schemaPath);
        } catch (_error) {
        }
      }

      for (const file of files) {
        await this.coverageService.addFromFile(file);
      }
    }
  }

  /** @type () => Promise<void> */
  async cleanAfterRun() {
    await this.coverageService.close();
    await fs.rm(this.coverageFilesDirectory, { recursive: true });

    // Remove empty reports directory, e.g. when only text-reporter is used
    if (readdirSync(this.options.reportsDirectory).length === 0) {
      await fs.rm(this.options.reportsDirectory, { recursive: true });
    }
  }

  /** @type CoverageProvider["reportCoverage"] */
  async reportCoverage(coverageMap) {
    this.generateReports(/** @type CoverageMap */ (coverageMap) ?? coverage.createCoverageMap());

    // In watch mode we need to preserve the previous results if cleanOnRerun is disabled
    const keepResults = !this.options.cleanOnRerun && this.ctx.config.watch;

    if (!keepResults) {
      await this.cleanAfterRun();
    }
  }

  /** @type (coverageMap: CoverageMap) => void */
  generateReports(coverageMap) {
    const context = libReport.createContext({
      dir: this.options.reportsDirectory,
      coverageMap
    });

    if (this.hasTerminalReporter(this.options.reporter)) {
      this.ctx.logger.log(c.blue(" % ") + c.dim("Coverage report from ") + c.yellow(this.name));
    }

    for (const reporter of this.options.reporter) {
      // Type assertion required for custom reporters
      reports
        .create(/** @type Parameters<typeof reports.create>[0] */ (reporter[0]), {
          projectRoot: this.ctx.config.root,
          ...reporter[1]
        })
        .execute(context);
    }
  }

  /** @type (reporters: ResolvedCoverageOptions["reporter"])=> boolean */
  hasTerminalReporter(reporters) {
    return reporters.some(([reporter]) => {
      return reporter === "text"
        || reporter === "text-summary"
        || reporter === "text-lcov"
        || reporter === "teamcity";
    });
  }

  /** @type CoverageProvider["onAfterSuiteRun"] */
  onAfterSuiteRun() {
    // The method is required by the interface, but doesn't seem to ever be called
    throw Error("Not Implemented");
  }

  /** @type CoverageProvider["generateCoverage"] */
  async generateCoverage() {
    const coverageMap = coverage.createCoverageMap();

    for (const file of await fs.readdir(this.coverageFilesDirectory, { recursive: true, withFileTypes: true })) {
      if (!file.isFile()) {
        continue;
      }

      const path = resolve(file.parentPath, file.name);
      /** @type CoverageMapData */
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const coverage = JSON.parse(await fs.readFile(path, "utf-8"));
      coverageMap.merge(coverage);
    }

    return coverageMap;
  }
}

/** @type (configReporters: NonNullable<BaseCoverageOptions["reporter"]>) => [string, Record<string, unknown>][] */
const resolveCoverageReporters = (configReporters) => {
  // E.g. { reporter: "html" }
  if (!Array.isArray(configReporters)) {
    return [[configReporters, {}]];
  }

  /** @type [string, Record<string, unknown>][] */
  const resolvedReporters = [];

  for (const reporter of configReporters) {
    if (Array.isArray(reporter)) {
      // E.g. { reporter: [ ["html", { skipEmpty: true }], ["lcov"], ["json", { file: "map.json" }] ]}
      resolvedReporters.push([reporter[0], /** @type Record<string, unknown> */ (reporter[1]) ?? {}]);
    } else {
      // E.g. { reporter: ["html", "json"]}
      resolvedReporters.push([reporter, {}]);
    }
  }

  return resolvedReporters;
};

export default JsonSchemaCoverageProviderModule;
