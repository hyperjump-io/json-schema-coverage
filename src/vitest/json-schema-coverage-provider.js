import { existsSync, readdirSync } from "node:fs";
import * as fs from "node:fs/promises";
import { createCoverageMap } from "istanbul-lib-coverage";
import { resolve } from "pathe";
import c from "tinyrainbow";
import libReport from "istanbul-lib-report";
import reports from "istanbul-reports";
import { coverageConfigDefaults } from "vitest/config";

/**
 * @import { BaseCoverageOptions, CoverageProvider, CoverageProviderModule, ResolvedCoverageOptions, Vitest} from "vitest"
 * @import { CoverageMap, CoverageMapData } from "istanbul-lib-coverage"
 */

/** @type CoverageProviderModule */
const JsonSchemaCoverageProviderModule = {
  /** @type CoverageProviderModule["getProvider"] */
  getProvider() {
    return new JsonSchemaCoverageProvider();
  }
};

/** @implements CoverageProvider */
class JsonSchemaCoverageProvider {
  name = "../src/vitest/json-schema-coverage-provider.js";

  ctx = /** @type Vitest */ ({});
  options = /** @type ResolvedCoverageOptions<"custom"> */ ({});
  coverageFilesDirectory = "";

  /** @type CoverageProvider["initialize"] */
  initialize(ctx) {
    this.ctx = ctx;

    const config = ctx.config.coverage;

    /** @type ResolvedCoverageOptions<"custom"> */
    this.options = {
      ...coverageConfigDefaults,

      // User's options
      ...config,

      // Resolved fields
      provider: "custom",
      customProviderModule: this.name,
      reportsDirectory: resolve(
        ctx.config.root,
        config.reportsDirectory || coverageConfigDefaults.reportsDirectory
      ),
      reporter: resolveCoverageReporters(config.reporter || coverageConfigDefaults.reporter),
      extension: ".schema.json"
    };

    this.coverageFilesDirectory = "./.json-schema-coverage";
  }

  /** @type CoverageProvider["resolveOptions"] */
  resolveOptions() {
    return /** @type NonNullable<any> */ (this.options);
  }

  /** @type CoverageProvider["clean"] */
  async clean(clean = true) {
    if (clean && existsSync(this.options.reportsDirectory)) {
      await fs.rm(this.options.reportsDirectory, {
        recursive: true,
        force: true,
        maxRetries: 10
      });
    }

    if (existsSync(this.coverageFilesDirectory)) {
      await fs.rm(this.coverageFilesDirectory, {
        recursive: true,
        force: true,
        maxRetries: 10
      });
    }

    await fs.mkdir(this.coverageFilesDirectory, { recursive: true });
  }

  /** @type () => Promise<void> */
  async cleanAfterRun() {
    await fs.rm(this.coverageFilesDirectory, { recursive: true });

    // Remove empty reports directory, e.g. when only text-reporter is used
    if (readdirSync(this.options.reportsDirectory).length === 0) {
      await fs.rm(this.options.reportsDirectory, { recursive: true });
    }
  }

  /** @type CoverageProvider["reportCoverage"] */
  async reportCoverage(coverageMap) {
    this.generateReports(/** @type CoverageMap */ (coverageMap) ?? this.createCoverageMap());

    // In watch mode we need to preserve the previous results if cleanOnRerun is disabled
    const keepResults = !this.options.cleanOnRerun && this.ctx.config.watch;

    if (!keepResults) {
      await this.cleanAfterRun();
    }
  }

  /** @type (coverageMap: CoverageMap) => void */
  generateReports(coverageMap) {
    const context = libReport.createContext({
      dir: this.options.reportsDirectory,
      coverageMap
    });

    if (this.hasTerminalReporter(this.options.reporter)) {
      this.ctx.logger.log(
        c.blue(" % ") + c.dim("Coverage report from ") + c.yellow(this.name)
      );
    }

    for (const reporter of this.options.reporter) {
      // Type assertion required for custom reporters
      reports
        .create(/** @type Parameters<typeof reports.create>[0] */ (reporter[0]), {
          projectRoot: this.ctx.config.root,
          ...reporter[1]
        })
        .execute(context);
    }
  }

  /** @type (reporters: ResolvedCoverageOptions["reporter"])=> boolean */
  hasTerminalReporter(reporters) {
    return reporters.some(
      ([reporter]) =>
        reporter === "text"
        || reporter === "text-summary"
        || reporter === "text-lcov"
        || reporter === "teamcity"
    );
  }

  /** @type () => CoverageMap */
  createCoverageMap() {
    return createCoverageMap({});
  }

  /** @type CoverageProvider["onAfterSuiteRun"] */
  onAfterSuiteRun() {
    // The method is required by the interface, but doesn't seem to ever be called
    throw Error("Not Implemented");
  }

  /** @type CoverageProvider["generateCoverage"] */
  async generateCoverage() {
    const coverageMap = this.createCoverageMap();

    for (const file of await fs.readdir(this.coverageFilesDirectory, { recursive: true, withFileTypes: true })) {
      const path = resolve(file.parentPath, file.name);
      /** @type CoverageMapData */
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const coverage = JSON.parse(await fs.readFile(path, "utf-8"));
      coverageMap.merge(coverage);
    }

    return coverageMap;
  }
}

/** @type (configReporters: NonNullable<BaseCoverageOptions["reporter"]>) => [string, Record<string, unknown>][] */
const resolveCoverageReporters = (configReporters) => {
  // E.g. { reporter: "html" }
  if (!Array.isArray(configReporters)) {
    return [[configReporters, {}]];
  }

  /** @type [string, Record<string, unknown>][] */
  const resolvedReporters = [];

  for (const reporter of configReporters) {
    if (Array.isArray(reporter)) {
      // E.g. { reporter: [ ["html", { skipEmpty: true }], ["lcov"], ["json", { file: "map.json" }] ]}
      resolvedReporters.push([reporter[0], /** @type Record<string, unknown> */ (reporter[1]) ?? {}]);
    } else {
      // E.g. { reporter: ["html", "json"]}
      resolvedReporters.push([reporter, {}]);
    }
  }

  return resolvedReporters;
};

export default JsonSchemaCoverageProviderModule;
