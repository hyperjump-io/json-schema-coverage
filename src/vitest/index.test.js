import { afterAll, afterEach, beforeAll, describe, expect, test } from "vitest";
import { mkdir, rm, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { registerSchema, unregisterSchema } from "./index.js";

describe("Worksheet", () => {
  test("matches with schema", async () => {
    await expect("foo").to.matchJsonSchema({ type: "string" });
  });

  test("doesn't match with schema", async () => {
    await expect(42).to.not.matchJsonSchema({ type: "string" });
  });

  test("match failure returns BASIC output", async () => {
    try {
      await expect(42).to.matchJsonSchema({
        $id: "https://example.com/main",
        type: "string"
      });
      expect.fail();
    } catch (error) {
      expect(/** @type Error */ (error).message).toMatchSnapshot();
    }
  });

  describe("with file-based schema in JSON", () => {
    const fixtureDirectory = ".vitest-matchers-json-fixtures";

    /** @type string */
    let fixturePath;

    /** @type (fixturePath: string, schemaJson: string) => Promise<void> */
    const addSchemaFile = async (fileName, schemaJson) => {
      fixturePath = resolve(fixtureDirectory, fileName);
      await writeFile(fixturePath, schemaJson);
    };

    beforeAll(async () => {
      await mkdir(fixtureDirectory, { recursive: true });
    });

    afterEach(async () => {
      await unregisterSchema(fixturePath);
      await rm(fixturePath);
    });

    afterAll(async () => {
      await rm(fixtureDirectory, { recursive: true });
    });

    test("match with registered schema's $id (json)", async () => {
      await addSchemaFile("subject.schema.json", `{
  "$id": "https://example.com/subject",
  "$schema": "https://json-schema.org/draft/2020-12/schema"
}`);
      await registerSchema(fixturePath);
      await expect(42).to.matchJsonSchema("https://example.com/subject");
    });

    test("match with schema's path (json)", async () => {
      await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema"
}`);
      await expect(42).to.matchJsonSchema(fixturePath);
    });

    test("match with registered schema's $id (yaml)", async () => {
      await addSchemaFile("subject.schema.yaml", `
$id: 'https://example.com/subject'
$schema: 'https://json-schema.org/draft/2020-12/schema'
`);
      await registerSchema(fixturePath);
      await expect(42).to.matchJsonSchema("https://example.com/subject");
    });

    test("match with schema's path (yaml)", async () => {
      await addSchemaFile("subject.schema.yaml", `
$schema: 'https://json-schema.org/draft/2020-12/schema'
`);
      await expect(42).to.matchJsonSchema(fixturePath);
    });

    test("match with registered schema's $id (yml)", async () => {
      await addSchemaFile("subject.schema.yml", `
$id: 'https://example.com/subject'
$schema: 'https://json-schema.org/draft/2020-12/schema'
`);
      await registerSchema(fixturePath);
      await expect(42).to.matchJsonSchema("https://example.com/subject");
    });

    test("match with schema's path (yml)", async () => {
      await addSchemaFile("subject.schema.yml", `
$schema: 'https://json-schema.org/draft/2020-12/schema'
`);
      await expect(42).to.matchJsonSchema(fixturePath);
    });

    test("registered with draft-04 id", async () => {
      await addSchemaFile("subject.schema.json", `{
  "id": "https://example.com/subject",
  "$schema": "http://json-schema.org/draft-04/schema#"
}`);
      await registerSchema(fixturePath);
      await expect(42).to.matchJsonSchema("https://example.com/subject");
    });
  });
});
