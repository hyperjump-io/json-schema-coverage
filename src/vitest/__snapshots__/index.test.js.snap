// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Worksheet > match failure returns BASIC output 1`] = `
"{
  "valid": false,
  "errors": [
    {
      "keyword": "https://json-schema.org/keyword/type",
      "absoluteKeywordLocation": "https://example.com/main#/type",
      "instanceLocation": "#"
    }
  ]
}"
`;
